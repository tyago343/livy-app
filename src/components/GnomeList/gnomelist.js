/** @jsxRuntime classic */
/** @jsx jsx */
import { css, jsx } from "@emotion/react";
import { useEffect, useState } from "react";
import { GnomeCard } from "../GnomeCard/gnomecard";
import { gnomesGrid, paginatorStyles } from "./gnomelist.styles";

const GnomeList = ({ gnomes, filter }) => {
  var [page, setPage] = useState(1);
  var [qtyPerPage] = useState(50);
  useEffect(() => {
    setPage(1);
  }, [filter]);
  const handleClick = (evt) => {
    setPage(Number(evt.target.id));
  };
  const indexOfLastGnome = page * qtyPerPage;
  const indexOfFirstGnome = indexOfLastGnome - qtyPerPage;
  const currentGnomes = gnomes?.slice(indexOfFirstGnome, indexOfLastGnome);
  const renderGnomes = currentGnomes?.map((gnome, index) => {
    return <GnomeCard gnome={gnome} key={index} />;
  });
  const pageNumbers = [];
  for (let i = 1; i <= Math.ceil(gnomes?.length / qtyPerPage); i++) {
    pageNumbers.push(i);
  }
  const renderPageNumbers = pageNumbers.map((number) => {
    if (number + 1 === page || number - 1 === page || number === page) {
      return (
        <li
          key={number}
          id={number}
          selected={number === page}
          onClick={handleClick}
          css={paginatorStyles}
        >
          {number}
        </li>
      );
    }
    return false;
  });
  return (
    <div data-testid="gnomelist">
      <h2>{`You're selected all the ${filter ? filter + "'s" : "workers"}`}</h2>
      <section css={gnomesGrid}>{renderGnomes}</section>
      <ol
        css={css`
          display: flex;
          justify-content: center;
          align-items: center;
          width: "100%";
        `}
        data-testid="paginator"
      >
        {renderPageNumbers}
      </ol>
    </div>
  );
};
export { GnomeList };
