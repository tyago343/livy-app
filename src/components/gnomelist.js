import { Link } from "react-router-dom";

const GnomeList = ({ gnomes }) => {
  return (
    <section>
      {gnomes.map((gnome, index) => (
        <Link to={`/gnome/${gnome.id}`} key={index}>
          <h2>{gnome.name}</h2>
          <img src={gnome.thumbnail} alt={gnome.name} srcset="" />
          <div>
            <h3>Characteristics:</h3>
            <ul>
              <li>
                Hair color:{" "}
                <span style={{ color: gnome.hair_color }}>
                  {gnome.hair_color}
                </span>
              </li>
              <li>
                Weight: <span>{parseFloat(gnome.weight).toFixed(2)}</span>
              </li>
              <li>
                Height: <span>{parseFloat(gnome.height).toFixed(2)}</span>
              </li>
              <li>
                Age: <span>{gnome.age}</span>
              </li>
            </ul>
          </div>
        </Link>
      ))}
    </section>
  );
};
export { GnomeList };
