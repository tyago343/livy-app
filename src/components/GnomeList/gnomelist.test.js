import { render, screen } from "@testing-library/react";
import { BrowserRouter } from "react-router-dom";
import { GnomeList } from "./gnomelist";
const props = {
  gnomes: [
    {
      id: 0,
      name: "Tobus Quickwhistle",
      thumbnail:
        "http://www.publicdomainpictures.net/pictures/10000/nahled/thinking-monkey-11282237747K8xB.jpg",
      age: 306,
      weight: 39.065952,
      height: 107.75835,
      hair_color: "Pink",
      professions: [
        "Metalworker",
        "Woodcarver",
        "Stonecarver",
        " Tinker",
        "Tailor",
        "Potter",
      ],
      friends: ["Cogwitz Chillwidget", "Tinadette Chillbuster"],
    },
    {
      id: 1,
      name: "Tobus Quickwhistle",
      thumbnail:
        "http://www.publicdomainpictures.net/pictures/10000/nahled/thinking-monkey-11282237747K8xB.jpg",
      age: 306,
      weight: 39.065952,
      height: 107.75835,
      hair_color: "Pink",
      professions: [
        "Metalworker",
        "Woodcarver",
        "Stonecarver",
        " Tinker",
        "Tailor",
        "Potter",
      ],
      friends: ["Cogwitz Chillwidget", "Tinadette Chillbuster"],
    },
  ],
};
describe("Test for the list of gnomes", () => {
  test("Gnome list display an array of gnomes", async () => {
    render(
      <BrowserRouter>
        <GnomeList {...props} />
      </BrowserRouter>
    );
    const items = await screen.findAllByTestId("gnome-card");
    expect(items).toHaveLength(props.gnomes.length);
  });
  test("Paginator is in the screen and contains qty of pages", async () => {
    render(
      <BrowserRouter>
        <GnomeList {...props} />
      </BrowserRouter>
    );
    expect(screen.getByTestId("paginator")).toMatchSnapshot(`
      <ol
        class="css-b58ksj-GnomeList"
        data-testid="paginator"
      >
        <li
          class="css-n1yxqf"
          id="1"
        >
          1
        </li>
      </ol>
    `);
  });
});
