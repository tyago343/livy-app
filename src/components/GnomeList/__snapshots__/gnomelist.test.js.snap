// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Test for the list of gnomes Paginator is in the screen and contains qty of pages: 
      <ol
        class="css-b58ksj-GnomeList"
        data-testid="paginator"
      >
        <li
          class="css-n1yxqf"
          id="1"
        >
          1
        </li>
      </ol>
     1`] = `
<ol
  class="css-b58ksj-GnomeList"
  data-testid="paginator"
>
  <li
    class="css-n1yxqf"
    id="1"
  >
    1
  </li>
</ol>
`;
